File: source/bindings/csharp/conversation_participants_changed_eventargs.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

using System;
using System.Collections.Generic;
using static Microsoft.CognitiveServices.Speech.Internal.ConversationTranslator;
using static Microsoft.CognitiveServices.Speech.Internal.SpxExceptionThrower;

namespace Microsoft.CognitiveServices.Speech.Transcription
{
    /// <summary>
    /// Why the participant changed event was raised
    /// Added in version 1.8.0
    /// </summary>
    public enum ParticipantChangedReason
    {
        /// <summary>
        /// Participant has joined the conversation.
        /// </summary>
        JoinedConversation,

        /// <summary>
        /// Participant has left the conversation. This could be voluntary, or involuntary
        /// (e.g. they are experiencing networking issues).
        /// </summary>
        LeftConversation,

        /// <summary>
        /// The participants' state has changed (e.g. they became muted, changed their nickname).
        /// </summary>
        Updated
    };

    /// <summary>
    /// Defines a payload for the <see cref="ConversationTranslator.ParticipantsChanged" /> event.
    /// </summary>
    public class ConversationParticipantsChangedEventArgs : SessionEventArgs
    {
        /// <summary>
        /// Creates a new instance.
        /// </summary>
        /// <param name="handle">The event handle.</param>
        internal ConversationParticipantsChangedEventArgs(IntPtr handle) : base(handle)
        {
            Reason = GetValue<ParticipantChangedReason>(handle, conversation_translator_event_get_participant_changed_reason);

            var participants = new List<Participant>();
            IntPtr participantHandle = IntPtr.Zero;

            try
            {
                for (int i = 0; participantHandle != SPXHANDLE_INVALID; i++)
                {
                    ThrowIfFail(conversation_translator_event_get_participant_changed_at_index(handle, i, out participantHandle));
                    if (participantHandle != SPXHANDLE_INVALID)
                    {
                        participants.Add(new Participant(participantHandle));
                        participantHandle = IntPtr.Zero;
                    }
                }
            }
            catch
            {
                // dispose of any created participants
                foreach (var p in participants)
                {
                    p.Dispose();
                }

                throw;
            }
            finally
            {
                if (participantHandle != SPXHANDLE_INVALID && participantHandle != IntPtr.Zero)
                {
                    ThrowIfFail(Internal.Participant.participant_release_handle(participantHandle));
                }
            }

            Participants = participants;
        }

        /// <summary>
        /// Why the participant changed event was raised (e.g. a participant joined).
        /// </summary>
        public ParticipantChangedReason Reason { get; }

        /// <summary>
        /// The participant(s) that joined, left, or were updated.
        /// </summary>
        public IReadOnlyList<Participant> Participants;
    }
}

File: source/bindings/java/com/microsoft/cognitiveservices/speech/transcription/ConversationParticipantsChangedEventArgs.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license
// information.
//
package com.microsoft.cognitiveservices.speech.transcription;

import com.microsoft.cognitiveservices.speech.SessionEventArgs;
import com.microsoft.cognitiveservices.speech.util.Contracts;
import com.microsoft.cognitiveservices.speech.util.IntRef;
import com.microsoft.cognitiveservices.speech.util.SafeHandle;
import java.util.ArrayList;
import java.util.List;

/** Defines a payload for the ParticipantsChanged event. Added in version 1.12.0 */
public class ConversationParticipantsChangedEventArgs extends SessionEventArgs {

    ConversationParticipantsChangedEventArgs(long eventArgs) {
        super(eventArgs);
        storeEventData(eventArgs, false);
    }

    ConversationParticipantsChangedEventArgs(long eventArgs, boolean dispose) {
        super(eventArgs);
        storeEventData(eventArgs, dispose);
    }

    /**
     * Why the participant changed event was raised (e.g. a participant joined).
     * @return the reason.
     */
    public ParticipantChangedReason getReason() {
        return this.reason;
    }

    /**
     * The participant(s) that joined, left, or were updated.
     * @return The participant(s).
     */
    public List<Participant> getParticipants() {
        return this.participants;
    }

    @Override
    public String toString() {
        return "SessionId:"
                + getSessionId()
                + " Reason:"
                + this.reason
                + " Participants:"
                + this.participants.size();
    }

    private void storeEventData(long eventArgs, boolean disposeNativeResources) {
        Contracts.throwIfNull(eventHandle, "eventHandle");

        IntRef intVal = new IntRef(0);
        Contracts.throwIfFail(getReason(eventHandle, intVal));
        this.reason = ParticipantChangedReason.values()[(int) intVal.getValue()];

        participants = new ArrayList<Participant>();

        IntRef resultRef = new IntRef(0);
        int index = 0;
        long result;
        result = getParticipantAt(eventHandle, index, resultRef);

        while (resultRef.getValue() > 0) {
            Participant participant = new Participant(resultRef.getValue());
            participants.add(participant);

            resultRef = new IntRef(0);
            index++;
            result = getParticipantAt(eventHandle, index, resultRef);
        }

        if (disposeNativeResources == true) {
            super.close();
        }
    }

    private List<Participant> participants;
    private ParticipantChangedReason reason;

    private final native long getParticipantAt(
            SafeHandle eventHandle, int index, IntRef participantRef);

    private final native long getReason(SafeHandle eventHandle, IntRef reasonVal);
}
